### PR TITLE
Issue #8577: Records Support check update for ClassMemberImpliedModifier

### PR DESCRIPTION
**Pull request by nmancus1**
Issue #8577: Records Support check update for ClassMemberImpliedModifier

Diff Regression projects: https://gist.githubusercontent.com/nmancus1/e3988f8092d919b5cbe70f5359c4d9e8/raw/a0f6f5860f8025c19868061dee6e0e40960b992f/projects-to-test-on.properties

Diff Regression config: https://gist.githubusercontent.com/nmancus1/862ab46a87aa3d4ffc29309420eac624/raw/8447a3976cfa3cadc15636fcd0d9e257e946796d/ClassMemberImpliedModifierBase.xml

### CODE REVIEW

**Comment by nmancus1** [comment]
Github, generate report

**Comment by nmancus1** [comment]
@esilkensen ping

**Comment by esilkensen** [comment]
@romani sorry, refreshing my memory: for these check update PR's, should we go through at least 3 reviews before merging, or OK with just me and @pbludov?

**Comment by esilkensen** [comment]
@nmancus1 bumping this PR, I wondered about the new NPE here:


https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/full-records-ClassMemberImpliedModifierCheck_2020210717/reports/diff/checkstyle/index.html#A28

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 1]
```diff
@@ -171,12 +203,13 @@ public void setViolateImpliedStaticOnNestedInterface(boolean violateImplied) {
...
             TokenTypes.INTERFACE_DEF,
             TokenTypes.ENUM_DEF,
+            TokenTypes.RECORD_DEF,
         };
     }
 
     @Override
     public void visitToken(DetailAST ast) {
-        if (ScopeUtil.isInClassBlock(ast) || ScopeUtil.isInEnumBlock(ast)) {
+        if (isInEnumClassOrRecordBlock(ast)) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:212]: `isInTypeBlock`

**Comment by nmancus1** [comment]
GitHub, generate report

**Comment by nmancus1** [comment]
@esilkensen sorry, I forgot to rerun the report. Please see the new one.

**Comment by esilkensen** [comment]
> @esilkensen sorry, I forgot to rerun the report. Please see the new one. 

Looks good, thanks 👍

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java (hunk 1 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:212]: @nmancus1, please do this renaming

Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java]: Done.

**Comment by pbludov** [comment]
> @romani sorry, refreshing my memory: for these check update PR's, should we go through at least 3 reviews before merging, or OK with just me and @pbludov?

I'm not @romani but the answer is "yes". Erik, the more reviewers, the better. Unfortunately, our resources are limited. We all have jobs and families. So, if possible, we need to rid @romani and other maintainers of trivial PRs. Usually, I look at the number of changes in the main code. If the changes affect only a couple of lines and all tests are green, there is not need to many reviewers.
For new modules and, especially, for changes in grammar, we need all possible reviewers. Anything that becomes public must be maintained. But it is very difficult to maintain something that is poorly designed, so in this case, calling a lot of reviewers can save resources in the future.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 2]
```diff
@@ -27,18 +27,18 @@
 
 /**
  * <p>
- * Checks for implicit modifiers on nested types in classes.
+ * Checks for implicit modifiers on nested types in classes and records.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:30]: if we treated enums and "nested type" I think we can treat record as "nested type" also.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 3]
```diff
@@ -27,18 +27,18 @@
...
  * </p>
  * <p>
  * This check is effectively the opposite of
  * <a href="https://checkstyle.org/config_modifier.html#RedundantModifier">RedundantModifier</a>.
- * It checks the modifiers on nested types in classes, ensuring that certain modifiers are
- * explicitly specified even though they are actually redundant.
+ * It checks the modifiers on nested types in classes and records, ensuring that certain modifiers
+ * are explicitly specified even though they are actually redundant.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:36]: same as above.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 4]
```diff
@@ -48,28 +48,34 @@
...
+ * Rationale for this check: Nested enums, interfaces, and records are treated differently from
+ * nested classes as they are only allowed to be {@code static}. Developers should not need to
+ * remember this rule, and this check provides the means to enforce that the modifier is coded
+ * explicitly.
  * </p>
  * <ul>
  * <li>
  * Property {@code violateImpliedStaticOnNestedEnum} - Control whether to enforce that
- * {@code static} is explicitly coded on nested enums in classes.
+ * {@code static} is explicitly coded on nested enums in classes and records.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:59]: why "records" are affecting violateImpliedStaticOnNestedEnum ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 5]
```diff
@@ -48,28 +48,34 @@
  * }
  * </pre>
  * <p>
- * Rationale for this check: Nested enums and interfaces are treated differently from nested
- * classes as they are only allowed to be {@code static}. Developers should not need to remember
- * this rule, and this check provides the means to enforce that the modifier is coded explicitly.
+ * Rationale for this check: Nested enums, interfaces, and records are treated differently from
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:51]: this is good extension, we should keep it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 6]
```diff
@@ -48,28 +48,34 @@
...
  * Type is {@code boolean}.
  * Default value is {@code true}.
  * </li>
  * </ul>
  * <p>
- * To configure the check so that it checks that all implicit modifiers on nested
- * interfaces and enums are
- * explicitly specified in classes.
+ * To configure the check so that it checks that all implicit modifiers on nested interfaces, enums,
+ * and records are explicitly specified in classes and records.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:78]: this is good extension, we should keep it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 7]
```diff
@@ -125,19 +139,25 @@
 
     /**
      * Control whether to enforce that {@code static} is explicitly coded
-     * on nested enums in classes.
+     * on nested enums in classes and records.
      */
     private boolean violateImpliedStaticOnNestedEnum = true;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:144]: why "record" affect this ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 8]
```diff
@@ -125,19 +139,25 @@
...
 
     /**
      * Control whether to enforce that {@code static} is explicitly coded
-     * on nested interfaces in classes.
+     * on nested interfaces in classes and records.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:148]: why "record" affect this ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java [hunk 9]
```diff
@@ -148,7 +168,7 @@ public void setViolateImpliedStaticOnNestedEnum(boolean violateImplied) {
 
     /**
      * Setter to control whether to enforce that {@code static} is explicitly coded
-     * on nested interfaces in classes.
+     * on nested interfaces in classes and records.
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:171]: why "record" affect this ?

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/classmemberimpliedmodifier/InputClassMemberImpliedModifierRecords.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    public static enum GoodEnum {} // OK
+    // Implied modifier 'static' should be explicit. [ClassMemberImpliedModifier]
+    public enum BadEnum {}
+
+    public static record GoodRecord() {} // OK
+    // Implied modifier 'static' should be explicit. [ClassMemberImpliedModifier]
+    public record BadRecord() {}
+
+    public static record OuterRecord() {
```
Comment (nmancus1) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/classmemberimpliedmodifier/InputClassMemberImpliedModifierRecords.java:21]: @romani we are checking for implied modifiers _inside_ of records, as well; this is why I changed the documentation to say "classes and records".

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java (hunk 8 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:148]: explained at https://github.com/checkstyle/checkstyle/pull/8578#discussion_r472288927

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java (hunk 9 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:171]: https://github.com/checkstyle/checkstyle/pull/8578#discussion_r472288927

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java (hunk 7 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:144]: https://github.com/checkstyle/checkstyle/pull/8578#discussion_r472288927

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java (hunk 4 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:59]: https://github.com/checkstyle/checkstyle/pull/8578#discussion_r472288927

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java (hunk 3 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:36]: https://github.com/checkstyle/checkstyle/pull/8578#discussion_r472288927

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java (hunk 2 above)
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/modifier/ClassMemberImpliedModifierCheck.java:30]: https://github.com/checkstyle/checkstyle/pull/8578#discussion_r472288927

**Pull request merged**